# Exploring Ebay Car Sales Data

This project is a `Data Cleaning Project` using a Kaggle dataset scraped from the German Ebay website, particularly in the 'Kleinanzeigen' section(a classified ads section in Ebay Germany).

NOTE: original dataset not in kaggle anymore. alternative link below.

For more info on dataset Documentation is [here](https://data.world/data-society/used-cars-data).

The aim of this project is to clean data and analyze used car data listings.

We will start by importing the necessary libraries.

In [2]:
import numpy as np
import pandas as pd

# Read csv file into pandas dataframe (utf-8 not working)
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
# use jupyter to render first few and last few values.
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
# print information about the autos dataframe
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
# show first 5 rows of the dataframe.
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Observations

- There are 50,000 rows and 20 columns in this dataset.
- Some Columns are not equal to 500,000 contains Null values. (none are > 20%)
- The Column names use `camel case` instead of python style `snake case`.
- Of the 20 columns, 15 are apparently of `string` datatype and 5 are of `int` datatype.
- Dataset loaded uses about 7.6 Megabytes of RAM.

# Data Cleaning

# Cleaning Column Names
Below we are going to convert the dataframe column names from camel case to snake case. We will also reword some column names to be more descriptive.

In [6]:
# print array of columns for overview
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# Cleaning column names to remove snake case and etc.
# Columns had to be manually cleaned due to camel casing 
# some colums also requires rephrasing.
autos_columns = autos.columns
autos_columns = autos_columns.str.replace('yearOfRegistration','registration_year')
autos_columns = autos_columns.str.replace('monthOfRegistration','registration_month')
autos_columns = autos_columns.str.replace('notRepairedDamage','unrepaired_damage')
autos_columns = autos_columns.str.replace('dateCreated','ad_created')
autos_columns = autos_columns.str.replace('dateCrawled','date_crawled')
autos_columns = autos_columns.str.replace('offerType','offer_type')
autos_columns = autos_columns.str.replace('vehicleType','vehicle_type')
autos_columns = autos_columns.str.replace('powerPS','power_ps')
autos_columns = autos_columns.str.replace('fuelType','fuel_type')
autos_columns = autos_columns.str.replace('nrOfPictures','nr_of_pictures')
autos_columns = autos_columns.str.replace('postalCode','postal_code')
autos_columns = autos_columns.str.replace('lastSeen','last_seen')

# reassign back to df.columns
autos.columns = autos_columns #cleaned version

# Print five samples to check current state of autos.
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The changes here are mainly focused on converting the previous camel cased column names to snake case which is the recommended case-style for python programs.

Other changes include rephrasing the column names to enhance the understandability of the Dataset. As we might observe, the changes were done in a manually keyed manner. This is mainly because the dataset has no uniform changes that can be applied and each name-change is unique.

snake-case conversion might have been done differently but we must first create a special function to achieve that as there are no standard modules that can achieve such. But that can be improved some other time since we have already achieved the column cleaning stage.

# Data Exploration and Cleaning

Below we will be doing some basic data exploration using the `df.describe()` method in order to get some descriptive statistics regarding the dataset.

In [8]:
# look at descriptive statistics for autos
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [9]:
# dataframe comparison column
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

- Odometer and Price seems to be missing the dataframe.describe() return value. This means that these values may be int and string combinations which require closer inspection.

In [10]:
# closer inspection on registration year
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
2013     806
2014     666
1994     660
2018     492
1993     445
2015     399
1990     395
1992     391
1991     356
1989     181
        ... 
1950       3
1955       2
9000       2
1954       2
1800       2
1957       2
1941       2
1951       2
1934       2
4100       1
4800       1
1953       1
1111       1
1927       1
6200       1
4500       1
1943       1
5911       1
1939       1
1938       1
2800       1
8888       1
1000       1
1500       1
1948       1
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

- On closer inspection of the registration year column we can see some years that are way beyond the reasonable range of years. year 9996 is way too far in the future and year 1000 may be to early for cars to exist, so these are points to check on in cleaning.

- in relation to the other columns, powerps and registration month may need some closer looks due to their zero values aligned with the items with year 1000 registration years.

below we shall start to clean the price and odometer columns.

# Exploring the Odometer and Price Columns

In [11]:
# price column inspect
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [12]:
# odometer column inspect
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

Our non-numeric value in these cases are the $ sign and km suffix. we shall start removing these and rename if necessary

In [13]:
# price column.
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

In [14]:
# odometer column
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)

# rename odometer column.
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

In [15]:
# inspect the cleaned versions of price and odometer
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In this step we will inspect the odometer_km column and look for outliers. 

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Based on preliminary assesments there seems to no noticeable problems in the odometer_km column that can be spotted so far.

We will proceed to the price column.

In [19]:
autos['price'].unique().shape

(2357,)

In [20]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [21]:
autos['price'].value_counts().sort_index(ascending = False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

The Price column contains about 1421 items that has prices of 0 dollars and a single item that costs 99999999 dollars these items are what we can say are outliers to the dataset so we must remove them.

In [22]:
# remove outliers using Boolean indexing.
x = 0
y = 99999999 

# series.between method not working properly may need some testing.
autos = autos[(autos['price'] > x) & (autos['price'] < y)]

In [23]:
# inspect cleaned column
autos['price'].value_counts().sort_index(ascending = False)

27322222      1
12345678      3
11111111      2
10000000      1
3890000       1
1300000       1
1234566       1
999999        2
999990        1
350000        1
345000        1
299000        1
295000        1
265000        1
259000        1
250000        1
220000        1
198000        1
197000        1
194000        1
190000        1
180000        1
175000        1
169999        1
169000        1
163991        1
163500        1
155000        1
151990        1
145000        1
           ... 
70           10
66            1
65            5
60            9
59            1
55            2
50           49
49            4
47            1
45            4
40            6
35            1
30            7
29            1
25            5
20            4
18            1
17            3
15            2
14            1
13            2
12            3
11            2
10            7
9             1
8             1
5             2
3             1
2             3
1           156
Name: price, Length: 235

We have succesfully removed the outliers.

# Exploring Date Columns

We shall now observe the date values contained in date_crawled, ad_created and last_seen.

In [24]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [25]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


The first ten characters in these columns are date formatted. We will extract these data and use them as a series to perform statistical analysis.

In [26]:
# store columns first as series objects for better ops
date_crawled = autos['date_crawled'].str[:10]
ad_created = autos['ad_created'].str[:10]
last_seen = autos['last_seen'].str[:10]

# start with date_crawled
date_crawled.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64

All postings are added by the crawler in 2016. Mostly in March and early April.

In [27]:
# ad_created
ad_created.value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033163
2016-03-10    0.031887
2016-03-11    0.032896
2016-03-12    0.036766
2016-03-13    0.017004
2016-03-14    0.035181
2016-03-15    0.034007
2016-03-16    0.030117
2016-03-17    0.031290
2016-03-18    0.013586
2016-03-19    0.033678
2016-03-20    0.037939
2016-03-21 

Ad postings span until back in 2015, but still the dataset limits observations just until april of 2016. so far everything seems reasonable.

In [28]:
# last_seen
last_seen.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: last_seen, dtype: float64

The last seen column still exhibits normal behavior. The dataset limits until 2016 April last seen data is still reasonable to have concluded that there are posts last seen until this date. No outliers can be detected.

In [29]:
# Let's try to observe the data inside registration_year
autos['registration_year'].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Based on our earlier observation we have detected anomalies with the registration year column wherein there are years 1000 and 9999. These are extreme outliers and will have to be removed.

To add more explanations, any registration beyond 2016 is inaccurate as we cannot see a listing of a car with a registration in the future.

# Dealing with incorrect Registration Year Data

We shall start fixing this column below

In [30]:
# We will have to determine what are the acceptable highest
# and lowest year values before we can proceed.
# We will detect upper and lower bounds first.

registration_year = autos['registration_year']
registration_year.value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      24
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2637
2002    2486
2003    2699
2004    2703
2005    2936
2006    2671
2007    2277
2008    2215
2009    2086
2010    1589
2011    1623
2012    1310
2013     803
2014     664
2015     392
2016    1220
2017    1393
2018     471
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

We can decide now that the most reasonable upper bound would be `2016`, anything beyond 2016 would be inaccurate data.

since cars were invented in 1886 it is impossible to see cars registered in 1800, we can then conclude the earliest possible date would be `1910`. This I believe is reasonable since antique ford vehicles do exist.

In [31]:
# remove the data points outside the upper and lower bounds.
x = 1910
y = 2016

autos = autos[(autos['registration_year'] >= x) & (autos['registration_year'] <= y)]

# check the registration_year column to see effects
autos['registration_year'].value_counts().sort_index()

1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      24
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
        ... 
1987      72
1988     135
1989     174
1990     347
1991     339
1992     371
1993     425
1994     629
1995    1228
1996    1373
1997    1951
1998    2363
1999    2898
2000    3156
2001    2637
2002    2486
2003    2699
2004    2703
2005    2936
2006    2671
2007    2277
2008    2215
2009    2086
2010    1589
2011    1623
2012    1310
2013     803
2014     664
2015     392
2016    1220
Name: registration_year, Length: 78, dtype: int64

Effects of the Boolean indexing is now effected and we have now removed irrelevant data from the registration year column.

# Exploring Price by Brand

In this section we will explore car brands in the dataset

In [32]:
autos_brand = autos['brand'].unique()
autos['brand'].value_counts().head(6)

volkswagen       9865
bmw              5138
opel             5023
mercedes_benz    4503
audi             4041
ford             3265
Name: brand, dtype: int64

The selected brands for aggregation are the top 6 brands with the most presence in the dataset.

In this portion we will compute the average price per brand.

In [33]:
selected_brands = autos['brand'].value_counts().head(6)
brand_ave_price = {}

for item in selected_brands.index:
    selected_brands = autos[autos['brand'] == item]
    ave_price = selected_brands['price'].mean().astype(int)
    brand_ave_price[item] = ave_price

In [34]:
print(brand_ave_price)

{'mercedes_benz': 8628, 'opel': 5432, 'audi': 9336, 'bmw': 8571, 'ford': 7456, 'volkswagen': 6729}


It looks like the most expensive car brand is audi, This is followed by bmw then mercedes-benz.

The results may vary from analysis to analysis due to the decisions to exclude or include certain datapoints in the cleaning process.

Sonstige_autos may seem like a brand but this isn't, This means other brands but in German.

# Storing Aggregate Data in a Dataframe

In [35]:
brand_ave_series = pd.Series(brand_ave_price)
brand_ave_df = pd.DataFrame(brand_ave_series, columns = ['mean_price'])

In [36]:
brand_ave_df.sort_values(['mean_price'], ascending = False)

mean_price
audi                 9336
mercedes_benz        8628
bmw                  8571
ford                 7456
volkswagen           6729
opel                 5432

We will now try to determine if there is a correlation between mileage and price for our average prices.

We will aggregate two series into a dataframe to perform a sort of correlation check between the 2 series' datapoints. We will use mean mileage for the second series.

In [37]:
selected_brands = autos['brand'].value_counts().head(6)
brand_ave_mileage = {}

for item in selected_brands.index:
    selected_brands = autos[autos['brand'] == item]
    ave_mileage = selected_brands['odometer_km'].mean().astype(int)
    brand_ave_mileage[item] = ave_mileage

In [38]:
print(brand_ave_mileage)

{'mercedes_benz': 130788, 'opel': 129314, 'audi': 129157, 'bmw': 132575, 'ford': 124243, 'volkswagen': 128713}


In [39]:
brand_mileage_series = pd.Series(brand_ave_mileage)
brand_ave_df['mean_mileage'] = brand_mileage_series

In [43]:
brand_ave_df.sort_values(['mean_price'], ascending = False)

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8571        132575
ford                 7456        124243
volkswagen           6729        128713
opel                 5432        129314

In [44]:
brand_ave_df.sort_values(['mean_mileage'], ascending = False)

mean_price  mean_mileage
bmw                  8571        132575
mercedes_benz        8628        130788
opel                 5432        129314
audi                 9336        129157
volkswagen           6729        128713
ford                 7456        124243

# Conclusion

Based on our analysis we can see that there is quite a weak correlation between the price and mileage data. A lower mileage count does not always correlate a higher price or vice versa.

Brand recognition and perceived manufacturing quality still plays a very big role in this analysis. We cannot quantify such concept as consumer psychology but we can infer in more advanced analysis like machine learning that certain brands may retain much more value than others.